File: lib/models/home_widget_snapshot.dart
/// ข้อมูลของวิดเจ็ตหน้าโฮมหนึ่งก้อน ตามที่เซิร์ฟเวอร์บอกมา
///
/// ข้อความไทยทุกบรรทัดในนี้ถูกเขียนมาแล้วจากฝั่ง Laravel (`HomeWidgetService`)
/// เหมือนการ์ดวันเดินทาง — แอปไม่แต่งประโยคเอง มีข้อยกเว้นข้อเดียวคือตัวเลข
/// "อีก N วัน" ที่ฝั่ง native นับใหม่จาก [HomeWidgetTrip.departureDate] ทุกครั้งที่
/// วาด เพราะวิดเจ็ตต้องนับถอยหลังถูกต้องข้ามคืนโดยไม่มีใครเปิดแอป
///
/// หน้าที่ของคลาสนี้คือ "ล้างชนิดข้อมูลให้เรียบร้อยก่อนข้ามไปฝั่ง native" —
/// PHP ส่ง 0 มาตรงที่ควรเป็น 0.0 ได้ ส่ง "3" มาตรงที่ควรเป็น 3 ได้ ถ้าปล่อยผ่านไป
/// ตรง ๆ ตัวถอดรหัสฝั่ง Swift/Kotlin จะพังทั้งก้อนแล้ววิดเจ็ตจะว่างเปล่าโดยไม่มี
/// ร่องรอยให้ตามหา
class HomeWidgetSnapshot {
  /// เวอร์ชันสัญญาที่ฝั่ง native ในบิลด์นี้อ่านได้ — ต้องตรงกับ
  /// `HomeWidgetService::SNAPSHOT_VERSION` ฝั่ง Laravel
  static const contractVersion = 1;

  /// ส่งต่อค่าที่เซิร์ฟเวอร์บอกมาตรง ๆ ไม่ทับด้วย [contractVersion] โดยตั้งใจ:
  /// ถ้าเซิร์ฟเวอร์ขึ้นเวอร์ชันก่อนที่แอปเครื่องนี้จะอัปเดต ฝั่ง native ต้องได้
  /// เห็นเลขจริงแล้วเลือกแสดงสถานะว่าง ดีกว่าวาดข้อมูลที่ตัวเองแปลผิด
  final int version;
  final String? generatedAt;
  final HomeWidgetTrip? trip;
  final HomeWidgetPayment? payment;

  const HomeWidgetSnapshot({
    required this.version,
    this.generatedAt,
    this.trip,
    this.payment,
  });

  static HomeWidgetSnapshot? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;

    return HomeWidgetSnapshot(
      version: _int(json['version']) ?? contractVersion,
      generatedAt: _string(json['generated_at']),
      trip: HomeWidgetTrip.fromJson(_map(json['trip'])),
      payment: HomeWidgetPayment.fromJson(_map(json['payment'])),
    );
  }

  bool get isEmpty => trip == null && payment == null;

  /// รูปแบบที่ฝั่ง native รอรับ — คีย์ชุดเดียวกับที่เซิร์ฟเวอร์ส่งมา (snake_case)
  /// เพื่อให้มีคำศัพท์ชุดเดียวตลอดสาย Laravel → Dart → Swift/Kotlin
  Map<String, dynamic> toNativeJson() => {
    'version': version,
    'generated_at': generatedAt,
    'trip': trip?.toNativeJson(),
    'payment': payment?.toNativeJson(),
  };
}

class HomeWidgetTrip {
  final String bookingRef;
  final String tripTitle;

  /// "2026-09-05" — ฝั่ง native นับ "อีกกี่วัน" ใหม่จากค่านี้
  final String? departureDate;

  /// วันสุดท้ายที่ก้อนนี้ยังจริง (วันกลับ) — ฝั่ง native เก็บการ์ดออกเองเมื่อเลยไปแล้ว
  final String? validUntil;
  final String dateLabel;
  final String? departTime;
  final int countdownDays;
  final String headline;
  final String detail;
  final String stage;
  final int? etaMinutes;
  final double progress;

  /// true = อยู่ในช่วงที่การ์ดหน้าจอล็อกทำงาน ให้เชื่อ [headline] ที่ส่งมาทั้งดุ้น
  /// false = ยังไม่ถึงวันเดินทาง ฝั่ง native นับวันเองได้
  final bool isLive;

  const HomeWidgetTrip({
    required this.bookingRef,
    required this.tripTitle,
    required this.dateLabel,
    required this.countdownDays,
    required this.headline,
    required this.detail,
    required this.stage,
    required this.progress,
    required this.isLive,
    this.departureDate,
    this.validUntil,
    this.departTime,
    this.etaMinutes,
  });

  static HomeWidgetTrip? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;

    final ref = _string(json['booking_ref']);
    // ไม่มีเลขการจองก็ไม่มีที่ให้แตะไป และแปลว่าก้อนนี้เพี้ยน — ทิ้งทั้งบล็อก
    if (ref == null) return null;

    return HomeWidgetTrip(
      bookingRef: ref,
      tripTitle: _string(json['trip_title']) ?? 'ทริปของคุณ',
      departureDate: _string(json['departure_date']),
      validUntil: _string(json['valid_until']),
      dateLabel: _string(json['date_label']) ?? '',
      departTime: _string(json['depart_time']),
      countdownDays: _int(json['countdown_days']) ?? 0,
      headline: _string(json['headline']) ?? 'ทริปของคุณ',
      detail: _string(json['detail']) ?? '',
      stage: _string(json['stage']) ?? 'countdown',
      etaMinutes: _int(json['eta_minutes']),
      progress: (_double(json['progress']) ?? 0).clamp(0.0, 1.0),
      isLive: json['is_live'] == true,
    );
  }

  Map<String, dynamic> toNativeJson() => {
    'booking_ref': bookingRef,
    'trip_title': tripTitle,
    'departure_date': departureDate,
    'valid_until': validUntil,
    'date_label': dateLabel,
    'depart_time': departTime,
    'countdown_days': countdownDays,
    'headline': headline,
    'detail': detail,
    'stage': stage,
    'eta_minutes': etaMinutes,
    'progress': progress,
    'is_live': isLive,
  };
}

class HomeWidgetPayment {
  final String bookingRef;
  final String tripTitle;
  final String label;
  final double amount;
  final String amountLabel;
  final String? dueDate;
  final String dueLabel;
  final int? daysLeft;
  final bool overdue;
  final bool slipPending;

  const HomeWidgetPayment({
    required this.bookingRef,
    required this.tripTitle,
    required this.label,
    required this.amount,
    required this.amountLabel,
    required this.dueLabel,
    required this.overdue,
    required this.slipPending,
    this.dueDate,
    this.daysLeft,
  });

  static HomeWidgetPayment? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;

    final ref = _string(json['booking_ref']);
    if (ref == null) return null;

    return HomeWidgetPayment(
      bookingRef: ref,
      tripTitle: _string(json['trip_title']) ?? 'ทริปของคุณ',
      label: _string(json['label']) ?? 'ยอดค้างชำระ',
      amount: _double(json['amount']) ?? 0,
      amountLabel: _string(json['amount_label']) ?? '',
      dueDate: _string(json['due_date']),
      dueLabel: _string(json['due_label']) ?? '',
      daysLeft: _int(json['days_left']),
      overdue: json['overdue'] == true,
      slipPending: json['slip_pending'] == true,
    );
  }

  Map<String, dynamic> toNativeJson() => {
    'booking_ref': bookingRef,
    'trip_title': tripTitle,
    'label': label,
    'amount': amount,
    'amount_label': amountLabel,
    'due_date': dueDate,
    'due_label': dueLabel,
    'days_left': daysLeft,
    'overdue': overdue,
    'slip_pending': slipPending,
  };
}

Map<String, dynamic>? _map(dynamic value) =>
    value is Map ? Map<String, dynamic>.from(value) : null;

/// สตริงว่างถือว่าไม่มีค่า — ฝั่ง native จะได้ไม่ต้องเช็ค `isEmpty` ทุกจุดที่วาด
String? _string(dynamic value) {
  if (value == null) return null;
  final text = value.toString().trim();
  return text.isEmpty ? null : text;
}

int? _int(dynamic value) {
  if (value is int) return value;
  if (value is num) return value.round();
  if (value == null) return null;
  return int.tryParse(value.toString());
}

double? _double(dynamic value) {
  if (value is double) return value;
  if (value is num) return value.toDouble();
  if (value == null) return null;
  return double.tryParse(value.toString());
}

File: ios/LiveActivity/HomeWidgetSnapshot.swift
import Foundation

/// สัญญาข้อมูลของวิดเจ็ตหน้าโฮม ระหว่างแอป ↔ วิดเจ็ต
///
/// ไฟล์นี้อยู่ในทั้งสองเป้าหมาย (Runner กับ LiveActivityExtension) โดยตั้งใจ — แอป
/// เป็นฝ่ายเขียน วิดเจ็ตเป็นฝ่ายอ่าน ทั้งคู่ต้องใช้ชื่อ App Group และคีย์เดียวกันเป๊ะ
/// ถ้าแยกเป็นสองไฟล์ วันหนึ่งจะพิมพ์ไม่ตรงกันแล้ววิดเจ็ตจะว่างเปล่าโดยไม่มี error
/// ให้เห็นเลย
///
/// ⚠️ ชื่อฟิลด์ต้องตรงกับที่ `HomeWidgetService` ฝั่ง Laravel ส่งมา (snake_case →
/// camelCase ด้วย `.convertFromSnakeCase`) ถ้าไม่ตรงแม้แต่ตัวเดียว การถอดรหัสจะล้ม
/// ทั้งก้อนแล้ววิดเจ็ตจะขึ้นสถานะว่าง
enum HomeWidgetStore {
  /// ต้องตรงกับที่ประกาศไว้ใน Runner.entitlements และ
  /// LiveActivityExtension.entitlements — และต้องเปิด App Group นี้ไว้ใน
  /// Apple Developer portal ด้วย ไม่งั้น `UserDefaults(suiteName:)` คืน nil เงียบ ๆ
  static let appGroup = "group.com.luilaykhao.app"

  static let snapshotKey = "home_widget_snapshot"

  /// ต้องตรงกับ `HomeWidgetService::SNAPSHOT_VERSION` ฝั่ง Laravel และ
  /// `HomeWidgetSnapshot.contractVersion` ฝั่ง Dart
  static let contractVersion = 1

  /// nil เมื่อ App Group ยังไม่ถูกเปิดให้บิลด์นี้ — ทุกผู้เรียกต้องรับ nil ได้
  /// โดยไม่ล้ม เพราะนี่คือสภาพของบิลด์ที่ profile ยังไม่มี capability
  private static var store: UserDefaults? {
    UserDefaults(suiteName: appGroup)
  }

  /// เขียน snapshot ก้อนใหม่ (JSON ดิบ ตามที่ฝั่ง Dart ประกอบมา)
  ///
  /// ตั้งใจไม่แกะ JSON ตรงนี้: ฝั่ง Dart ล้างชนิดข้อมูลมาแล้ว และการแกะซ้ำที่นี่
  /// หมายถึงมีสองที่ที่ต้องแก้ทุกครั้งที่สัญญาเปลี่ยน
  @discardableResult
  static func write(json: String) -> Bool {
    guard let store else { return false }
    store.set(json, forKey: snapshotKey)
    return true
  }

  @discardableResult
  static func clear() -> Bool {
    guard let store else { return false }
    store.removeObject(forKey: snapshotKey)
    return true
  }

  /// snapshot ล่าสุดที่แอปเขียนไว้ — nil เมื่อยังไม่มี, อ่านไม่ได้, หรือคนละเวอร์ชัน
  static func read() -> HomeWidgetSnapshot? {
    guard let json = store?.string(forKey: snapshotKey) else { return nil }
    return decode(json)
  }

  /// แยกออกมาจาก [read] เพื่อให้ทดสอบได้โดยไม่ต้องมี App Group
  ///
  /// นี่คือจุดที่สัญญาระหว่าง Dart กับ Swift พังได้เงียบที่สุด — ชื่อฟิลด์เพี้ยนตัวเดียว
  /// ก็คืน nil ทั้งก้อนแล้ววิดเจ็ตขึ้นสถานะว่างโดยไม่มี error ให้เห็น
  /// (`ios/scripts/check_home_widget_contract.swift` ยิง JSON จริงจากฝั่ง Dart เข้ามาที่นี่)
  static func decode(_ json: String) -> HomeWidgetSnapshot? {
    guard let data = json.data(using: .utf8) else { return nil }

    let decoder = JSONDecoder()
    decoder.keyDecodingStrategy = .convertFromSnakeCase

    guard let snapshot = try? decoder.decode(HomeWidgetSnapshot.self, from: data) else {
      return nil
    }

    // เวอร์ชันที่บิลด์นี้ไม่รู้จัก = วาดไปก็เสี่ยงแปลผิด ยอมขึ้นสถานะว่างดีกว่า
    guard snapshot.version == contractVersion else { return nil }

    return snapshot
  }
}

struct HomeWidgetSnapshot: Codable {
  let version: Int
  let generatedAt: String?
  let trip: HomeWidgetTrip?
  let payment: HomeWidgetPayment?

  var isEmpty: Bool { trip == nil && payment == nil }
}

struct HomeWidgetTrip: Codable {
  let bookingRef: String
  let tripTitle: String

  /// "2026-09-05" — วิดเจ็ตนับ "อีกกี่วัน" ใหม่จากค่านี้ทุกครั้งที่วาด
  let departureDate: String?

  /// วันกลับ — เลยวันนี้ไปแล้วถือว่าก้อนนี้หมดอายุ วิดเจ็ตเก็บการ์ดออกเอง
  let validUntil: String?

  let dateLabel: String
  let departTime: String?
  let countdownDays: Int
  let headline: String
  let detail: String
  let stage: String
  let etaMinutes: Int?
  let progress: Double

  /// true = อยู่ในช่วงที่การ์ดหน้าจอล็อกทำงาน ให้ใช้ [headline] ที่เซิร์ฟเวอร์เขียนมา
  /// ทั้งดุ้น (มันรู้เรื่องรถและ ETA ซึ่งวิดเจ็ตไม่รู้)
  let isLive: Bool
}

struct HomeWidgetPayment: Codable {
  let bookingRef: String
  let tripTitle: String
  let label: String
  let amount: Double
  let amountLabel: String
  let dueDate: String?
  let dueLabel: String
  let daysLeft: Int?
  let overdue: Bool
  let slipPending: Bool
}

/// การนับวันของฝั่งวิดเจ็ต
///
/// นี่คือที่เดียวที่ยอมให้ฝั่ง native คิดเลขเอง เพราะวิดเจ็ตต้องนับถอยหลังถูกต้อง
/// ข้ามคืนโดยไม่มีเน็ตและไม่มีใครเปิดแอป (ถ้ารอค่าจากเซิร์ฟเวอร์ คนที่ไม่เปิดแอป
/// สามวันจะเห็น "อีก 17 วัน" ค้างอยู่ทั้งสามวัน)
///
/// ⚠️ ปฏิทินและ locale ต้องกำหนดเองทั้งคู่ เครื่องคนไทยจำนวนมากตั้งปฏิทินเป็นพุทธ
/// ศักราช ถ้าใช้ค่าเริ่มต้นของเครื่อง DateFormatter จะอ่าน "2026-09-05" เป็นปี พ.ศ.
/// แล้วนับถอยหลังผิดไปห้าร้อยกว่าปี
enum HomeWidgetClock {
  static let bangkok: TimeZone =
    TimeZone(identifier: "Asia/Bangkok") ?? TimeZone(secondsFromGMT: 7 * 3600) ?? .current

  static var calendar: Calendar {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = bangkok
    return calendar
  }

  private static let dayParser: DateFormatter = {
    let formatter = DateFormatter()
    formatter.calendar = Calendar(identifier: .gregorian)
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = bangkok
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter
  }()

  static func day(from string: String?) -> Date? {
    guard let string, !string.isEmpty else { return nil }
    return dayParser.date(from: string)
  }

  /// จำนวนวันเต็มจาก "วันนี้ที่กรุงเทพ" ถึงวันเป้าหมาย (ติดลบ = ผ่านไปแล้ว)
  static func daysUntil(_ dateString: String?, from now: Date = Date()) -> Int? {
    guard let target = day(from: dateString) else { return nil }
    let calendar = self.calendar

    return calendar.dateComponents(
      [.day],
      from: calendar.startOfDay(for: now),
      to: calendar.startOfDay(for: target)
    ).day
  }

  /// เที่ยงคืนถัดไปตามเวลาไทย — จุดที่ตัวเลขนับถอยหลังต้องเปลี่ยน
  static func nextMidnight(after date: Date) -> Date? {
    calendar.nextDate(
      after: date,
      matching: DateComponents(hour: 0, minute: 0, second: 0),
      matchingPolicy: .nextTime
    )
  }
}
